motionplan/plannerOptions.go

```diff
@@ -156,3 +163,156 @@ func (p *plannerOptions) SetMaxSolutions(maxSolutions int) {
+		topoConstraints = true
+		p.addPbOrientationConstraints(from, to, orientationConstraint)
+	}
+	return topoConstraints
+}
+
+func (p *plannerOptions) addPbLinearConstraints(from, to spatialmath.Pose, pbConstraint *pb.LinearConstraint) {
+	// Linear constraints
+	linTol := pbConstraint.GetLineToleranceMm()
+	if linTol == 0 {
+		// Default
+		linTol = defaultLinearDeviation
+	}
+	orientTol := pbConstraint.GetOrientationToleranceDegs()
+	if orientTol == 0 {
+		orientTol = defaultOrientationDeviation
+	}
+	constraint, pathDist := NewAbsoluteLinearInterpolatingConstraint(from, to, float64(linTol), float64(orientTol))
+	p.AddConstraint(defaultLinearConstraintName, constraint)
+
+	// TODO(pl): check whether
+	// ~ p.pathDist = CombineMetrics(p.pathDist, pathDist)
+	p.pathDist = pathDist
+}
+
+func (p *plannerOptions) addPbOrientationConstraints(from, to spatialmath.Pose, pbConstraint *pb.OrientationConstraint) {
+	orientTol := pbConstraint.GetOrientationToleranceDegs()
+	if orientTol == 0 {
+		orientTol = defaultOrientationDeviation
+	}
+	constraint, pathDist := NewSlerpOrientationConstraint(from, to, float64(orientTol))
+	p.AddConstraint(defaultLinearConstraintName, constraint)
+	p.pathDist = CombineMetrics(p.pathDist, pathDist)
+}
+
+func (p *plannerOptions) createCollisionConstraints(
+	frame referenceframe.Frame,
+	fs referenceframe.FrameSystem,
+	worldState *referenceframe.WorldState,
+	seedMap map[string][]referenceframe.Input,
+	pbConstraint []*pb.CollisionSpecification,
+	reportDistances bool,
+	logger golog.Logger,
+) error {
+	allowedCollisions := []*Collision{}
+
+	// List of geometries which may be specified for collision ignoring
+	validGeoms := map[string]bool{}
+
+	addGeomNames := func(geomsInFrame *referenceframe.GeometriesInFrame) error {
+		for geomName := range geomsInFrame.Geometries() {
+			if _, ok := validGeoms[geomName]; ok {
+				return fmt.Errorf("geometry %s is specified by name more than once", geomName)
+			}
+			validGeoms[geomName] = true
+		}
+		return nil
+	}
+
+	// Get names of world state obstacles
+	if worldState != nil {
+		for _, geomsInFrame := range worldState.Obstacles {
+			err := addGeomNames(geomsInFrame)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// TODO(pl): non-moving frame system geometries are not currently supported for collision avoidance ( RSDK-2129 ) but are included here
+	// in anticipation of support and to prevent spurious errors.
+	allFsGeoms, err := referenceframe.FrameSystemGeometries(fs, seedMap, logger)
+	if err != nil {
+		return err
+	}
+	for _, geomsInFrame := range allFsGeoms {
+		err = addGeomNames(geomsInFrame)
+		if err != nil {
+			return err
+		}
+	}
+
+	// This allows the user to specify an entire component with sub-geometries, e.g. "myUR5arm", and the specification will apply to all
+	// sub-pieces, e.g. myUR5arm:upper_arm_link, myUR5arm:base_link, etc. Individual sub-pieces may also be so addressed.
+	allowNameToSubGeoms := func(cName string) ([]string, error) {
```

Comment:
I love this idea.

Reply:
Does this have an impact if someone specifies the `myUR5arm` should not collide but the `myUR5arm:ee_link` should collide?